Makes a bunch of things material-respecting

comma opened the pull request:
Rods, lattices, grilles and windows are now material based. Different materials affect looks and health mostly.
Can make rods of arbitrary materials now, as long as they're hard enough (50 hardness or more).
Can reinforce sheet stacks with rods of other materials. Currently only does something when building windows from resulting stacks. Can get reinforcement material back with a welder.
90% changes are just repathing rods to material stack

Fixes #10526.

PsiOmegaDelta commented on code/__defines/materials.dm:
```diff
@@ -70,3 +70,8 @@
+//Stack flags
+#define USE_MATERIAL_COLOR 			(1 << 0)
+#define USE_MATERIAL_SINGULAR_NAME    (1 << 1)
+#define USE_MATERIAL_PLURAL_NAME    	(1 << 2)
```
Travis hates bit shifting. You should probably use the hex values that are already in use above.

comma replied:
ree

PsiOmegaDelta commented on code/game/objects/items/stacks/sheets/glass.dm:
```diff
@@ -14,17 +13,47 @@
+
+/obj/item/stack/material/glass/Initialize(mapload, var/new_material, var/new_reinf_material)
+	if(!set_material(new_material, new_reinf_material))
+		return INITIALIZE_HINT_QDEL
```
Must call `..()` even if you're returning a different hint.

PsiOmegaDelta commented on code/game/objects/structures/grille.dm:
```diff
@@ -3,28 +3,45 @@
+/obj/structure/grille/Initialize(mapload, var/new_material)
+	material = SSmaterials.get_material_by_name(new_material)
+	if(!istype(material))
+		return INITIALIZE_HINT_QDEL
```
Must call `..()` even if you're returning a different qdel hint.

PsiOmegaDelta commented on code/game/objects/structures/window.dm:
```diff
@@ -9,28 +9,78 @@
+		new_material = MATERIAL_GLASS
+	material = SSmaterials.get_material_by_name(new_material)
+	if(!istype(material))
+		return INITIALIZE_HINT_QDEL
```
Must call `..()` even if you're returning a different qdel hint.

comma replied:
Calling it now first thing

comma committed:
Makes a bunch of things material-respecting

Rods, lattices, grilles and windows are now material based. Different materials affect looks and health mostly.
Can make rods of arbitrary materials now, as long as they're hard enough (50 hardness or more).
Can reinforce sheet stacks with rods of other materials. Currently only does something when building windows from resulting stacks. Can get reinforcement material back with a welder.
90% changes are just repathing rods to material stack

comma commented:
Will resolve map conflicts after code review, it's simple search & replace for rod paths.

afterthought2 commented on code/game/objects/structures/grille.dm:
```diff
@@ -3,28 +3,47 @@
+
+/obj/structure/grille/New(var/newloc, var/new_material)
+	..()
 	update_connections(1)
```
Does this really not work in `Initialize`?

afterthought2 commented on code/game/objects/structures/grille.dm:
```diff
@@ -119,13 +138,13 @@
 
 	src.health -= damage*0.2
-	spawn(0) healthcheck() //spawn to make sure we return properly if the grille is deleted
+	healthcheck() //spawn to make sure we return properly if the grille is deleted
```
Kill the comment.

afterthought2 commented on code/game/objects/structures/grille.dm:
```diff
@@ -247,34 +240,46 @@
-		..()
-		health = rand(-5, -1) //In the destroyed but not utterly threshold.
-		healthcheck() //Send this to healthcheck just in case we want to do something else with it.
+/obj/structure/grille/broken/New()
```
Make `Initialize` for sanity.

afterthought2 commented on code/game/objects/structures/grille.dm:
```diff
@@ -247,34 +240,46 @@
+	to_chat(usr, "<span class='notice'>You assemble a grille</span>")
+	ST.in_use = 0
+	F.add_fingerprint(usr)
+	ST.use(2)
```
Should probably make sure it worked before spawning the object.

afterthought2 commented on code/game/objects/structures/window.dm:
```diff
@@ -9,28 +9,78 @@
+/obj/structure/window/get_material()
+	return material
+
+/obj/structure/window/New(var/newloc, start_dir=null, constructed=0, var/new_material, var/new_reinf_material)
```
I don't get it; why add both, and then change the material twice? ideally everything would be in Initialize.

afterthought2 commented on code/game/objects/structures/window.dm:
```diff
@@ -222,47 +230,50 @@
 			qdel(src)
+	else if(polarized && isMultitool(W))
+		var/t = sanitizeSafe(input(user, "Enter the ID for the window.", src.name, null), MAX_NAME_LEN)
+		if(user.incapacitated() && !user.Adjacent(src))
```
Being incapacitated is fine, provided you are adjacent?

afterthought2 commented on code/game/objects/structures/window.dm:
```diff
@@ -222,47 +230,50 @@
+			return
+		if (user.get_active_hand() != W)
+			return
+		if (!in_range(src, user) && src.loc != user)
```
Shouldn't the adjacent check handle this?

afterthought2 commented on code/game/objects/structures/window.dm:
```diff
@@ -222,47 +230,50 @@
+			return
+		if (!in_range(src, user) && src.loc != user)
+			return
+		t = sanitizeSafe(t, MAX_NAME_LEN)
```
It's being double-sanitized here.

afterthought2 commented on code/game/turfs/turf_changing.dm:
```diff
@@ -1,7 +1,7 @@
-/turf/proc/ReplaceWithLattice()
+/turf/proc/ReplaceWithLattice(var/material)
 	src.ChangeTurf(get_base_turf_by_area(src))
 	spawn()
```
Why is this being spawned?

comma replied:
no idea, predates me

afterthought2 commented on code/modules/materials/_materials.dm:
```diff
@@ -122,18 +121,21 @@
-		return
+/material/proc/reinforce(var/mob/user, var/obj/item/stack/material/used_stack, var/obj/item/stack/material/target_stack)
 	if(used_stack.get_amount() < 1 || target_stack.get_amount() < 1)
 		to_chat(user, "<span class='warning'>You need one rod and one sheet of [display_name] to make anything useful.</span>")
```
Maybe I'm not following, but it's not actually relevant that the stack is a rod? Along similar lines, shouldn't this proc check the material amount per stack var to determine how much to use?

comma replied:
actually yeah might as well make it general sheet interaction

afterthought2 commented on code/modules/materials/definitions/materials_glass.dm:
```diff
@@ -38,6 +37,9 @@
+	if(!used_stack.reinf_material)
+		options -= "Windoor"
 	var/title = "Sheet-[used_stack.name] ([used_stack.get_amount()] sheet\s left)"
 	var/choice = input(title, "What would you like to construct?") as null|anything in window_options
```
You use `window_options` here, allowing someone to select windoor when they shouldn't.

afterthought2 reviewed (changes requested):
Concerns:

- Something seems to not work with windoors checking the reinforcement of the material, and you just can't build them, getting the "not sufficiently reinforced" message.
- If you fail to build the window/windoor/whatever you select, you are thrown into the general crafting menu, and can then select things like glass beds to build. There is no other way to access this menu, as far as I can tell, so if you _want_ to build a glass bed, you have to build/fail to build a window first.

Will test some more once stuff is addressed.

comma commented:
Gonna do something about windows, probably turn them into proper stack recipies.
Also will make reinforcing a general sheet interaction.

comma committed:
Reinforced material recipies

Moved windows from special handling to stack recipies.
Made recipies for reinforced materials. May god forgive me.

comma commented:
Okay made them stack recipies. Reinforcing stuff is now general sheet on sheet interaction (1 reinforcing to 2 base sheets)

afterthought2 commented on code/modules/materials/material_sheets.dm:
```diff
@@ -54,53 +57,89 @@
+		if(is_same(W))
+			..()
+		else if(!reinf_material)
+			material.reinforce(user, W, src)
```
Certain "material" stacks are improperly initialized to an incorrect material, e.g. corgi skin is material steel. If one such stack is attacked by a "valid" reinforcing material, apparently the result is that it's deleted entirely (not clear on reason/details, can try it out).

afterthought2 commented on code/modules/materials/material_sheets.dm:
```diff
@@ -54,53 +57,89 @@
 	if(M) M.update_strings()
 	return transfer
 
-/obj/item/stack/material/attack_self(var/mob/user)
```
Not clear on the details, but currently building reinfoced stuff doesn't work. Probably the proc called by this thing's parent, `list_recipes`, doesn't correctly parse the new recipe var's format. In any case, windows are built but not reinforced, while windoors are unavailable.

babydoll commented:
Re-open once you're back into the swing of things mr chinsky
